metaflow/plugins/kfp/tests/.gitlab-ci.yml

```diff
@@ -1,211 +1,142 @@
+
+  if [ "${CI_COMMIT_BRANCH}" = "${CI_DEFAULT_BRANCH}" ]; then
+    PY_LIBRARY_VERSION="${METAFLOW_VERSION}.${CI_PIPELINE_IID}"
+    IMAGE_TAG="${METAFLOW_VERSION}.${CI_PIPELINE_IID}"
+  else
+    PY_LIBRARY_VERSION="0.0.${CI_PIPELINE_IID}+${CI_COMMIT_SHORT_SHA}"
+    IMAGE_TAG="0.0.${CI_PIPELINE_IID}"
+  fi
+
+.test:
+  extends: .generate_kubeconfig
+  stage: test
   variables:
     GIT_STRATEGY: none
     PIPELINE_VERSION: "1.0"
-  stage: test
   script:
-    - *path-to-host-file-directory
-    - export BUILT_IMAGE_FULL_PATH=$( cat ${_IMAGE_TAG_FILE} ) # reading the path to the Docker image on Artifactory
+    - *path-to-host-file-directory  
+    - export BUILT_IMAGE_FULL_PATH=$( cat ${IMAGE_REPOSITORY_TAG_PATH} )
     - cp -r /root/.kube .
     - docker run
         --rm
         -v ${SHARED_PATH}/.kube:/home/zservice/.kube
-        -v ${SHARED_PATH}/public:/home/zservice/public 
+        -v ${SHARED_PATH}/public:/home/zservice/public  
         -e AWS_ACCESS_KEY_ID=$AWS_ACCESS_KEY_ID 
         -e AWS_SECRET_ACCESS_KEY=$AWS_SECRET_ACCESS_KEY 
         -e AWS_SESSION_TOKEN=$AWS_SESSION_TOKEN 
         -e AWS_DEFAULT_REGION=$AWS_DEFAULT_REGION 
+        -e KFP_RUN_URL_PREFIX=$KFP_RUN_URL_PREFIX 
+        -e KFP_SDK_NAMESPACE=$KFP_SDK_NAMESPACE 
+        -e USER=$GITLAB_USER_EMAIL
         ${BUILT_IMAGE_FULL_PATH}
         bash -c "
-          export KFP_RUN_URL_PREFIX=https://kubeflow.corp.dev-k8s.zg-aip.net/ && 
-          export KFP_SDK_NAMESPACE=metaflow-integration-testing-internal && 
-          export USER=${GITLAB_USER_EMAIL} &&
           cd /home/zservice/metaflow/metaflow/plugins/kfp/tests &&
           python -m pytest -s -n 3 run_integration_tests.py --image ${BUILT_IMAGE_FULL_PATH} --opsgenie-api-token ${OPSGENIE_API_TOKEN} --cov-config=setup.cfg
         "
   artifacts:
     when: always
     paths:
       - public
+  
+
+build:docker:
+  stage: build
+  script:
+    - *version
+    - export IMAGE_REPOSITORY_TAG=${DOCKER_REPO_URL}/${CI_PROJECT_NAMESPACE}/${PY_LIBRARY_NAME}:${IMAGE_TAG}
+    - 'docker build --no-cache -f metaflow/plugins/kfp/tests/Dockerfile -t ${IMAGE_REPOSITORY_TAG} .'
+    - 'echo ${DOCKER_API_KEY} | docker login -u ${DOCKER_USERNAME} --password-stdin ${DOCKER_REPO_URL}'
+    - STDERR=$(docker push ${IMAGE_REPOSITORY_TAG} 2>&1 >/dev/null) || true
+    - echo $STDERR
+    - |
+      if [ -z "$STDERR" ] || [[ $STDERR == *"manifest invalid"* ]]; then
+        echo "Successful!"
+      else
+        echo "Error in pushing image to Artifactory."
+        exit 1
+      fi
+    - echo ${IMAGE_REPOSITORY_TAG} > ${IMAGE_REPOSITORY_TAG_PATH}
+  artifacts:
+    paths:
+      - ${IMAGE_REPOSITORY_TAG_PATH}
+
+build:publish:
+  extends: .aip_python_debian_image
+  stage: build
+  script:
+  - *version
+  # Now write back the zillow-kfp version back to the original location we found the original.
+  - sed -i "s/\(version = ['\"]\)[^'\"]*\(['\"]\)/\1${PY_LIBRARY_VERSION}\2/" $METAFLOW_VERSION_PATH
```

Comment:
Curious about this bash magic, it looks like you're replacing the second occurrence of `PY_LIBRARY_VERSION` with the first `zillow-kfp` version. Why is that? Unless I'm mistaking the reason for the `1` and `2` values in the expression.

Reply:
Effectively it's because when you run `pip list | grep metaflow` you want that version to be the version you expect. We auto-generate the version number using the Gitlab Pipeline ID so we don't have to automatically set it (as it's a pain to auto-set things!).

What this regex does is take a line like `version = '1.2.3'` and rewrite it as `version = '{py_library_version}'`. The `\1` is `version = '` and the `\2` is `'` 😄

Reply:
Makes sense, thanks for the explanation!